File: frontend/appflowy_flutter/lib/plugins/database_view/application/row/row_service.dart
import 'package:dartz/dartz.dart';
import 'package:appflowy_backend/dispatch/dispatch.dart';
import 'package:appflowy_backend/protobuf/flowy-error/errors.pb.dart';
import 'package:appflowy_backend/protobuf/flowy-database2/row_entities.pb.dart';

typedef RowId = String;

class RowBackendService {
  final String viewId;

  RowBackendService({
    required this.viewId,
  });

  Future<Either<RowMetaPB, FlowyError>> createRowAfterRow(RowId rowId) {
    final payload = CreateRowPayloadPB.create()
      ..viewId = viewId
      ..startRowId = rowId;

    return DatabaseEventCreateRow(payload).send();
  }

  Future<Either<OptionalRowPB, FlowyError>> getRow(RowId rowId) {
    final payload = RowIdPB.create()
      ..viewId = viewId
      ..rowId = rowId;

    return DatabaseEventGetRow(payload).send();
  }

  Future<Either<RowMetaPB, FlowyError>> getRowMeta(RowId rowId) {
    final payload = RowIdPB.create()
      ..viewId = viewId
      ..rowId = rowId;

    return DatabaseEventGetRowMeta(payload).send();
  }

  Future<Either<Unit, FlowyError>> updateMeta({
    required String rowId,
    String? iconURL,
    String? coverURL,
    bool? isDocumentEmpty,
  }) {
    final payload = UpdateRowMetaChangesetPB.create()
      ..viewId = viewId
      ..id = rowId;

    if (iconURL != null) {
      payload.iconUrl = iconURL;
    }
    if (coverURL != null) {
      payload.coverUrl = coverURL;
    }

    if (isDocumentEmpty != null) {
      payload.isDocumentEmpty = isDocumentEmpty;
    }

    return DatabaseEventUpdateRowMeta(payload).send();
  }

  static Future<Either<Unit, FlowyError>> deleteRow(
    String viewId,
    RowId rowId,
  ) {
    final payload = RowIdPB.create()
      ..viewId = viewId
      ..rowId = rowId;

    return DatabaseEventDeleteRow(payload).send();
  }

  static Future<Either<Unit, FlowyError>> duplicateRow(
    String viewId,
    RowId rowId, [
    String? groupId,
  ]) {
    final payload = RowIdPB.create()
      ..viewId = viewId
      ..rowId = rowId;
    if (groupId != null) {
      payload.groupId = groupId;
    }

    return DatabaseEventDuplicateRow(payload).send();
  }
}

File: frontend/appflowy_tauri/src/appflowy_app/application/database/row/row_service.ts
import {
  CreateRowPayloadPB,
  MoveGroupRowPayloadPB,
  MoveRowPayloadPB,
  OrderObjectPositionTypePB,
  RowIdPB,
  UpdateRowMetaChangesetPB,
} from '@/services/backend';
import {
  DatabaseEventCreateRow,
  DatabaseEventDeleteRow,
  DatabaseEventDuplicateRow,
  DatabaseEventGetRowMeta,
  DatabaseEventMoveGroupRow,
  DatabaseEventMoveRow,
  DatabaseEventUpdateRowMeta,
} from '@/services/backend/events/flowy-database2';
import { pbToRowMeta, RowMeta } from './row_types';

export async function createRow(viewId: string, params?: {
  position?: OrderObjectPositionTypePB;
  rowId?: string;
  groupId?: string;
  data?: Record<string, string>;
}): Promise<RowMeta> {
  const payload = CreateRowPayloadPB.fromObject({
    view_id: viewId,
    row_position: {
      position: params?.position,
      object_id: params?.rowId,
    },
    group_id: params?.groupId,
    data: params?.data,
  });

  const result = await DatabaseEventCreateRow(payload);

  return result.map(pbToRowMeta).unwrap();
}

export async function duplicateRow(viewId: string, rowId: string, groupId?: string): Promise<void> {
  const payload = RowIdPB.fromObject({
    view_id: viewId,
    row_id: rowId,
    group_id: groupId,
  });

  const result = await DatabaseEventDuplicateRow(payload);

  return result.unwrap();
}

export async function deleteRow(viewId: string, rowId: string, groupId?: string): Promise<void> {
  const payload = RowIdPB.fromObject({
    view_id: viewId,
    row_id: rowId,
    group_id: groupId,
  });

  const result = await DatabaseEventDeleteRow(payload);

  return result.unwrap();
}

export async function moveRow(viewId: string, fromRowId: string, toRowId: string): Promise<void> {
  const payload = MoveRowPayloadPB.fromObject({
    view_id: viewId,
    from_row_id: fromRowId,
    to_row_id: toRowId,
  });

  const result = await DatabaseEventMoveRow(payload);

  return result.unwrap();
}

/**
 * Move the row from one group to another group
 *
 * @param fromRowId
 * @param toGroupId
 * @param toRowId used to locate the moving row location.
 * @returns
 */
export async function moveGroupRow(viewId: string, fromRowId: string, toGroupId: string, toRowId?: string): Promise<void> {
  const payload = MoveGroupRowPayloadPB.fromObject({
    view_id: viewId,
    from_row_id: fromRowId,
    to_group_id: toGroupId,
    to_row_id: toRowId,
  });

  const result = await DatabaseEventMoveGroupRow(payload);

  return result.unwrap();
}


export async function getRowMeta(viewId: string, rowId: string, groupId?: string): Promise<RowMeta> {
  const payload = RowIdPB.fromObject({
    view_id: viewId,
    row_id: rowId,
    group_id: groupId,
  });

  const result = await DatabaseEventGetRowMeta(payload);

  return result.map(pbToRowMeta).unwrap();
}

export async function updateRowMeta(
  viewId: string,
  rowId: string,
  meta: {
    iconUrl?: string;
    coverUrl?: string;
  },
): Promise<void> {
  const payload = UpdateRowMetaChangesetPB.fromObject({
    view_id: viewId,
    id: rowId,
    icon_url: meta.iconUrl,
    cover_url: meta.coverUrl,
  });

  const result = await DatabaseEventUpdateRowMeta(payload);

  return result.unwrap();
}
